core/src/main/scala/kafka/coordinator/group/GroupCoordinatorAdapter.scala

```diff
@@ -234,4 +240,67 @@ class GroupCoordinatorAdapter(
     }
     CompletableFuture.completedFuture(results)
   }
+
+  override def commitTransactionalOffsets(
+    context: RequestContext,
+    request: TxnOffsetCommitRequestData,
+    bufferSupplier: BufferSupplier
+  ): CompletableFuture[TxnOffsetCommitResponseData] = {
+    val future = new CompletableFuture[TxnOffsetCommitResponseData]()
+
+    def callback(results: Map[TopicPartition, Errors]): Unit = {
+      val response = new TxnOffsetCommitResponseData()
+      val byTopics = new util.HashMap[String, TxnOffsetCommitResponseData.TxnOffsetCommitResponseTopic]()
+
+      results.forKeyValue { (tp, error) =>
+        var topic = byTopics.get(tp.topic)
+        if (topic == null) {
+          topic = new TxnOffsetCommitResponseData.TxnOffsetCommitResponseTopic().setName(tp.topic)
+          byTopics.put(tp.topic, topic)
+          response.topics.add(topic)
+        }
+        topic.partitions.add(new TxnOffsetCommitResponseData.TxnOffsetCommitResponsePartition()
+          .setPartitionIndex(tp.partition)
+          .setErrorCode(error.code))
+      }
+
+      future.complete(response)
+    }
+
+    val currentTimestamp = time.milliseconds
+    val partitions = new mutable.HashMap[TopicPartition, OffsetAndMetadata]()
+
+    request.topics.forEach { topic =>
+      topic.partitions.forEach { partition =>
+        val tp = new TopicPartition(topic.name, partition.partitionIndex)
+        partitions += tp -> new OffsetAndMetadata(
+          offset = partition.committedOffset,
+          leaderEpoch = partition.committedLeaderEpoch match {
+            case RecordBatch.NO_PARTITION_LEADER_EPOCH => Optional.empty[Integer]
```

Comment:
Did we previously not handle -1 correctly here?

Reply:
It was handled [here](https://github.com/apache/kafka/blob/a8d54231632355493b9d6e24431227199b44daea/clients/src/main/java/org/apache/kafka/common/requests/TxnOffsetCommitRequest.java#L124) in the previous code.

Reply:
Oh didn't see this. Thanks for sharing 😅 I like that both are handled in the same place now.